src/osd/PG.cc

```diff
@@ -4547,39 +4547,47 @@ void PG::share_pg_info()
 
 void PG::append_log_entries_update_missing(
   const list<pg_log_entry_t> &entries,
+  bool handle_missing,
   ObjectStore::Transaction &t)
 {
   assert(!entries.empty());
   assert(entries.begin()->version > info.last_update);
 
-  PGLogEntryHandler rollbacker;
-  pg_log.append_new_log_entries(
-    info.last_backfill,
-    info.last_backfill_bitwise,
-    entries,
-    &rollbacker);
-  rollbacker.apply(this, &t);
-  info.last_update = pg_log.get_head();
+  if (handle_missing) {
+    PGLogEntryHandler rollbacker;
+    pg_log.append_new_log_entries(info.last_backfill,
+				  info.last_backfill_bitwise,
+				  entries,
+				  &rollbacker);
+    rollbacker.apply(this, &t);
+    info.last_update = pg_log.get_head();
+    info.stats.stats_invalid = true;
+  } else {
+    for (auto &entry : entries) {
+      pg_log.add(entry);
+    }
+    info.last_update = pg_log.get_head();
+  }
 
   if (pg_log.get_missing().num_missing() == 0) {
     // advance last_complete since nothing else is missing!
     info.last_complete = info.last_update;
   }
 
-  info.stats.stats_invalid = true;
   dirty_info = true;
   write_if_dirty(t);
 }
 
 
 void PG::merge_new_log_entries(
   const list<pg_log_entry_t> &entries,
+  bool handle_missing,
   ObjectStore::Transaction &t)
 {
   dout(10) << __func__ << " " << entries << dendl;
   assert(is_primary());
 
-  append_log_entries_update_missing(entries, t);
+  append_log_entries_update_missing(entries, handle_missing, t);
```

Comment:
It seems like append_log_entries could indicate whether the log requires the missing set to be rebuilt based on the log entries?